tests/unit/utils/test_http.py

```diff
@@ -43,6 +43,7 @@ class TestIsSafeUrl:
             "\njavascript:alert(x)",
             "\x08//example.com",
             "\n",
+            "view/?param=//example.com",
```

Comment:
I guess there's no harm in considering _more_ URLs to be unsafe, but the change here is somewhat curious.

Reply:
Indeed it is - it's a difference between stdlib and urllib3 handling the same string.

```pycon
>>> from urllib.parse import urlparse
>>> from urllib3.util import parse_url
>>> url = "view/?param=//example.com"
>>> std_lib = urlparse(url)
>>> new_hotness = parse_url(url)
>>> repr(std_lib)
ParseResult(scheme='', netloc='', path='view/', params='', query='param=//example.com', fragment='')
>>> repr(new_hotness)
Url(scheme=None, auth=None, host='view', port=None, path='/', query='param=//example.com', fragment=None)
```

We can see that the parser now places the `host` (aka `netloc`) as `view`. Arguably the correct intended behavior?

The new parser trips up the final evaluation here:

https://github.com/pypi/warehouse/blob/72fbc6f9ce90db0a332616062aaabf2508e6d190/warehouse/utils/http.py#L50

Since the test suite is explicitly passing a host value of "testserver", the conditions no longer match.

@di do you think this is cause for concern?

cc @sethmlarson - is this parsing difference expected, and correct?

Reply:
I defer to @sethmlarson 🙂

Reply:
This is "expected" by urllib3 standards because we "handle" scheme-less URLs (which I dislike very much, and [we've deprecated in the latest version](https://github.com/urllib3/urllib3/pull/2950)). In a future version of urllib3 that URL will parse identically (and correctly according to RFC 3986) to urlparse.

For now the best way to protect against this is to assert a truthy scheme and host are parsed anywhere you're expecting an absolute URL.

Reply:
Thanks @sethmlarson ! It'll be some time before we're on the 2.x series and will handle the deprecation warning when we have it (which might mean moving the test example back to the original "allowed").
For now, I think we're good.